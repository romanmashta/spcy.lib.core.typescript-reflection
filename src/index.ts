import _ from 'lodash';
import ts, { NodeArray, SourceFile, TypeElement } from 'typescript';
import {
  EnumDeclaration,
  InterfaceDeclaration,
  MetaInfo,
  Module,
  TypeInfo,
  TypeLiteral,
  TypeReference
} from './meta-data';

const defaultOptions: ts.CompilerOptions = {
  declaration: false
};

interface NamedInfo<T> {
  [name: string]: T;
}

class MetaGenerator {
  private files: string[];
  private options: ts.CompilerOptions;
  private program: ts.Program;
  private sources: SourceFile[];
  private typeChecker: ts.TypeChecker;

  constructor(files: string[], options: ts.CompilerOptions = {}) {
    this.files = files;
    this.options = options;
    this.program = ts.createProgram(files, { ...defaultOptions, ...options });
    this.sources = _.map(this.files, f => this.program.getSourceFile(f)!);
    this.typeChecker = this.program.getTypeChecker();
  }

  inspectProperty = (node: ts.PropertySignature): NamedInfo<TypeInfo> => {
    const name = (node.name as ts.Identifier).text;
    const info: TypeInfo = this.inspectType(node.type!);
    return { [name]: info };
  };

  processMembers = (members: NodeArray<TypeElement>): TypeLiteral => ({
    properties: _.chain(members)
      .filter(ts.isPropertySignature)
      .reduce((r: any, prop: ts.PropertySignature) => ({ ...r, ...this.inspectProperty(prop) }), {})
      .value()
  });

  inspectTypeLiteral = (node: ts.TypeLiteralNode): TypeLiteral => {
    return { ...this.processMembers(node.members) };
  };

  inspectTypeRef = (node: ts.TypeReferenceNode): TypeReference => {
    const typeRef = (node.typeName as ts.Identifier).text;
    const args = node.typeArguments ? _.map(node.typeArguments, this.inspectType) : undefined;
    return { typeRef, arguments: args };
  };

  inspectType = (node: ts.TypeNode): TypeInfo => {
    switch (node.kind) {
      case ts.SyntaxKind.StringKeyword:
        return 'string';
      case ts.SyntaxKind.NumberKeyword:
        return 'number';
      case ts.SyntaxKind.BooleanKeyword:
        return 'boolean';
      case ts.SyntaxKind.TypeReference:
        return this.inspectTypeRef(node as ts.TypeReferenceNode);
      case ts.SyntaxKind.TypeLiteral:
        return this.inspectTypeLiteral(node as ts.TypeLiteralNode);
      default:
        return 'string';
    }
  };

  inspectInterface = (node: ts.InterfaceDeclaration): NamedInfo<InterfaceDeclaration> => {
    const name = node.name.text;
    const info: InterfaceDeclaration = {
      interface: {
        ...this.processMembers(node.members)
      }
    };
    return { [name]: info };
  };

  inspectEnumMember = (node: ts.EnumMember): NamedInfo<string> => {
    const name = (node.name as ts.Identifier).text;
    return { [name]: name };
  };

  inspectEnum = (node: ts.EnumDeclaration): NamedInfo<EnumDeclaration> => {
    const name = node.name.text;
    const info: EnumDeclaration = {
      enum: _.chain(node.members)
        .filter(ts.isEnumMember)
        .reduce((r: any, prop: ts.EnumMember) => ({ ...r, ...this.inspectEnumMember(prop) }), {})
        .value()
    };
    return { [name]: info };
  };

  transform = (): MetaInfo => {
    const metaInfo: MetaInfo = {
      modules: [],
      hasErrors: false
    };

    _.forEach(this.sources, sourceFile => {
      const module: Module = { members: {} };
      metaInfo.modules = [...metaInfo.modules, module];

      const inspect = (node: ts.Node) => {
        if (ts.isInterfaceDeclaration(node)) {
          module.members = { ...module.members, ...this.inspectInterface(node) };
        } else if (ts.isEnumDeclaration(node)) {
          module.members = { ...module.members, ...this.inspectEnum(node) };
        } else ts.forEachChild(node, inspect);
      };

      inspect(sourceFile);
    });

    return metaInfo;
  };
}

export const generateMetaInfoForFiles = (files: string[], options: ts.CompilerOptions): MetaInfo =>
  new MetaGenerator(files, options).transform();

export const generateMetaInfoForFile = (file: string, options: ts.CompilerOptions = {}): MetaInfo =>
  new MetaGenerator([file], options).transform();
