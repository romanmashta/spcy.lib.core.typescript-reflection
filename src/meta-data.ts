export type BasicTypes = 'boolean' | 'number' | 'string';

interface TypeReference {
  typeRef: string;
  arguments?: TypeInfo[];
}

type TypeInfo = BasicTypes | TypeReference;

interface InterfaceDeclaration {
  interface: {
    extends?: TypeInfo;
    properties: {
      [name: string]: TypeInfo;
    };
  };
}

interface EnumDeclaration {
  enum: {
    [name: string]: string | number;
  };
}

interface ImportDeclaration {
  import: string;
}

export interface Module {
  [name: string]: ImportDeclaration | InterfaceDeclaration | EnumDeclaration;
}
