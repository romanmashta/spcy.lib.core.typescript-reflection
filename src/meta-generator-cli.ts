import _ from 'lodash';
import * as ts from 'typescript';
import path from 'path';
import fs from 'fs';
import minimatch from 'minimatch';
import handlebars from 'handlebars';
import { SourceFile, TypeInfo } from '@spcy/lib.core.reflection';
import { generateMetaInfoForFiles } from './meta-generator';
import { ModuleTemplate } from './templates';

const readConfigFromFile = (configFileName: string): ts.ParsedCommandLine => {
  const config = ts.sys.readFile(configFileName);
  if (!config) throw new Error(`Cannot read config file: ${configFileName}`);
  const result = ts.parseConfigFileTextToJson(configFileName, config);
  const configObject = result.config;

  return ts.parseJsonConfigFileContent(
    configObject,
    ts.sys,
    path.dirname(configFileName),
    {},
    path.basename(configFileName)
  );
};

handlebars.registerHelper(
  'stringify',
  (object: TypeInfo) => new handlebars.SafeString(JSON.stringify(object, undefined, 4))
);

const renderModule = handlebars.compile(ModuleTemplate);

const writeSchemaFile = (sourceFile: SourceFile) => {
  const schemaFileName = sourceFile.fileName.replace(/model\.ts$/i, 'schema.ts');
  const moduleText = renderModule(sourceFile);
  fs.writeFile(schemaFileName, moduleText, (err: Error | null) => {
    if (err) {
      throw new Error(`Unable to write schema file: ${err.message}`);
    }
  });
};

export const run = (configFileName?: string) => {
  const resolvedConfigFileName = configFileName || path.join(process.cwd(), 'tsconfig.json');
  if (!fs.existsSync(resolvedConfigFileName)) {
    throw new Error(`Cannot find tsconfig.json ${resolvedConfigFileName}`);
  }
  const config = readConfigFromFile(resolvedConfigFileName);
  const modelFiles = config.fileNames.filter(minimatch.filter('*.model.ts', { matchBase: true }));
  const result = generateMetaInfoForFiles(modelFiles, config.options);
  _.forEach(result.sourceFiles, writeSchemaFile);
};
