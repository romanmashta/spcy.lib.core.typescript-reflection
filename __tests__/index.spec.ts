import * as path from 'path';
import * as _ from 'lodash';
import * as fs from 'fs';
import { generateMetaInfoForFile, exec } from '../src';

const CASES_ROOT = '__tests__/cases';
const MODULE_PATH = '__tests__/module';

const assertSchema = (caseName: string) => {
  const file = path.resolve(`${CASES_ROOT}/${caseName}/index.ts`);
  const metaFile = path.resolve(`${CASES_ROOT}/${caseName}/meta-data.ts`);
  const result = generateMetaInfoForFile(file);
  const module = _.first(result.modules);
  const { meta } = require(metaFile);

  //console.log(JSON.stringify(module, null, 2));
  expect(module).toEqual(meta);

  expect(result.hasErrors).toBe(false);
};

const caseNames = [
  'index-signature',
  'array',
  'type-alias',
  'one-of',
  'basic-interface',
  'enum',
  'mixed-types',
  'basic-typeref',
  'typeliteral',
  'typeliteral-argument',
  'meta-schema',
  'required-properties'
];

it.each(caseNames)('Process schema %s', caseName => {
  assertSchema(caseName);
});

it('process module', () => {
  const cwd = process.cwd();
  const files = exec({ path: path.join(cwd, MODULE_PATH) });
  files.forEach(f => expect(fs.existsSync(f)).toBe(true));
});
