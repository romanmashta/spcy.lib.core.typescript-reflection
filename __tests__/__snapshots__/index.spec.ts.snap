// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process schema array 1`] = `
Object {
  "$defs": Object {
    "Config": Object {
      "$id": "#/$defs/Config",
      "additionalProperties": undefined,
      "properties": Object {
        "colors": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "params": Object {
          "items": Object {
            "additionalProperties": undefined,
            "properties": Object {
              "key": Object {
                "type": "string",
              },
              "value": Object {
                "type": "string",
              },
            },
            "required": Array [
              "key",
              "value",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "required": Array [
        "colors",
        "params",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema basic-interface 1`] = `
Object {
  "$defs": Object {
    "Person": Object {
      "$id": "#/$defs/Person",
      "additionalProperties": undefined,
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "firstName": Object {
          "type": "string",
        },
        "isActive": Object {
          "type": "boolean",
        },
        "lastName": Object {
          "type": "string",
        },
      },
      "required": Array [
        "firstName",
        "lastName",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema basic-typeref 1`] = `
Object {
  "$defs": Object {
    "Location": Object {
      "$id": "#/$defs/Location",
      "additionalProperties": undefined,
      "properties": Object {
        "lat": Object {
          "type": "number",
        },
        "lon": Object {
          "type": "number",
        },
      },
      "required": Array [
        "lat",
        "lon",
      ],
      "type": "object",
    },
    "Person": Object {
      "$id": "#/$defs/Person",
      "additionalProperties": undefined,
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "firstName": Object {
          "type": "string",
        },
        "isActive": Object {
          "type": "boolean",
        },
        "lastName": Object {
          "type": "string",
        },
        "location": Object {
          "$ref": "#/$defs/Location",
        },
        "role": Object {
          "$ref": "#/$defs/Role",
        },
      },
      "required": Array [
        "firstName",
        "lastName",
        "age",
        "isActive",
        "location",
        "role",
      ],
      "type": "object",
    },
    "Role": Object {
      "$id": "#/$defs/Role",
      "oneOf": Array [
        Object {
          "const": "Guest",
        },
        Object {
          "const": "User",
        },
        Object {
          "const": "Admin",
        },
      ],
    },
  },
}
`;

exports[`Process schema enum 1`] = `
Object {
  "$defs": Object {
    "Role": Object {
      "$id": "#/$defs/Role",
      "oneOf": Array [
        Object {
          "const": "Guest",
        },
        Object {
          "const": "User",
        },
        Object {
          "const": "Admin",
        },
      ],
    },
  },
}
`;

exports[`Process schema index-signature 1`] = `
Object {
  "$defs": Object {
    "Config": Object {
      "$id": "#/$defs/Config",
      "additionalProperties": Object {
        "$ref": "#/$defs/Section",
      },
      "properties": undefined,
      "required": undefined,
      "type": "object",
    },
    "Section": Object {
      "$id": "#/$defs/Section",
      "additionalProperties": undefined,
      "properties": Object {
        "id": Object {
          "type": "string",
        },
        "secret": Object {
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "secret",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema inheritance 1`] = `
Object {
  "$defs": Object {
    "Audit": Object {
      "$id": "#/$defs/Audit",
      "additionalProperties": undefined,
      "properties": Object {
        "createdOn": Object {
          "type": "string",
        },
        "updatedOn": Object {
          "type": "string",
        },
      },
      "required": Array [
        "createdOn",
        "updatedOn",
      ],
      "type": "object",
    },
    "Entity": Object {
      "$id": "#/$defs/Entity",
      "additionalProperties": undefined,
      "properties": Object {
        "id": Object {
          "type": "string",
        },
      },
      "required": Array [
        "id",
      ],
      "type": "object",
    },
    "Person": Object {
      "$id": "#/$defs/Person",
      "allOf": Array [
        Object {
          "$ref": "#/$defs/Audit",
        },
        Object {
          "$ref": "#/$defs/Entity",
        },
        Object {
          "additionalProperties": undefined,
          "properties": Object {
            "age": Object {
              "type": "number",
            },
            "firstName": Object {
              "type": "string",
            },
            "isActive": Object {
              "type": "boolean",
            },
            "lastName": Object {
              "type": "string",
            },
          },
          "required": Array [
            "firstName",
            "lastName",
          ],
          "type": "object",
        },
      ],
    },
  },
}
`;

exports[`Process schema meta-schema 1`] = `
Object {
  "$defs": Object {
    "AllOf": Object {
      "$id": "#/$defs/AllOf",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "allOf": Object {
          "items": Object {
            "$ref": "#/$defs/TypeInfo",
          },
          "type": "array",
        },
      },
      "required": Array [
        "allOf",
      ],
      "type": "object",
    },
    "ArrayType": Object {
      "$id": "#/$defs/ArrayType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "items": Object {
          "$ref": "#/$defs/TypeInfo",
        },
        "maxItems": Object {
          "type": "number",
        },
        "minItems": Object {
          "type": "number",
        },
        "type": Object {
          "const": "array",
        },
        "uniqueItems": Object {
          "type": "boolean",
        },
      },
      "required": Array [
        "type",
        "items",
      ],
      "type": "object",
    },
    "BooleanType": Object {
      "$id": "#/$defs/BooleanType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "type": Object {
          "const": "boolean",
        },
      },
      "required": Array [
        "type",
      ],
      "type": "object",
    },
    "ConstLiteral": Object {
      "$id": "#/$defs/ConstLiteral",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "const": Object {
          "oneOf": Array [
            Object {
              "type": "string",
            },
            Object {
              "type": "number",
            },
            Object {
              "type": "boolean",
            },
            Object {
              "type": "null",
            },
          ],
        },
      },
      "required": Array [
        "const",
      ],
      "type": "object",
    },
    "DateType": Object {
      "$id": "#/$defs/DateType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "type": Object {
          "const": "date",
        },
      },
      "required": Array [
        "type",
      ],
      "type": "object",
    },
    "EnumType": Object {
      "$id": "#/$defs/EnumType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "enum": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
      },
      "required": Array [
        "enum",
      ],
      "type": "object",
    },
    "MetaInfo": Object {
      "$id": "#/$defs/MetaInfo",
      "additionalProperties": undefined,
      "properties": Object {
        "hasErrors": Object {
          "type": "boolean",
        },
        "modules": Object {
          "items": Object {
            "$ref": "#/$defs/Module",
          },
          "type": "array",
        },
        "sourceFiles": Object {
          "items": Object {
            "$ref": "#/$defs/SourceFile",
          },
          "type": "array",
        },
      },
      "required": Array [
        "sourceFiles",
        "modules",
        "hasErrors",
      ],
      "type": "object",
    },
    "Module": Object {
      "$id": "#/$defs/Module",
      "additionalProperties": undefined,
      "properties": Object {
        "$defs": Object {
          "additionalProperties": Object {
            "$ref": "#/$defs/TypeInfo",
          },
          "properties": undefined,
          "required": undefined,
          "type": "object",
        },
      },
      "required": Array [
        "$defs",
      ],
      "type": "object",
    },
    "NullType": Object {
      "$id": "#/$defs/NullType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "type": Object {
          "const": "null",
        },
      },
      "required": Array [
        "type",
      ],
      "type": "object",
    },
    "NumberType": Object {
      "$id": "#/$defs/NumberType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "maximum": Object {
          "type": "number",
        },
        "minimum": Object {
          "type": "number",
        },
        "type": Object {
          "const": "number",
        },
      },
      "required": Array [
        "type",
      ],
      "type": "object",
    },
    "ObjectType": Object {
      "$id": "#/$defs/ObjectType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "additionalProperties": Object {
          "oneOf": Array [
            Object {
              "$ref": "#/$defs/TypeInfo",
            },
            Object {
              "type": "boolean",
            },
          ],
        },
        "properties": Object {
          "additionalProperties": Object {
            "$ref": "#/$defs/TypeInfo",
          },
          "properties": undefined,
          "required": undefined,
          "type": "object",
        },
        "required": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "type": Object {
          "const": "object",
        },
      },
      "required": Array [
        "type",
      ],
      "type": "object",
    },
    "OneOf": Object {
      "$id": "#/$defs/OneOf",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "oneOf": Object {
          "items": Object {
            "$ref": "#/$defs/TypeInfo",
          },
          "type": "array",
        },
      },
      "required": Array [
        "oneOf",
      ],
      "type": "object",
    },
    "SourceFile": Object {
      "$id": "#/$defs/SourceFile",
      "additionalProperties": undefined,
      "properties": Object {
        "fileName": Object {
          "type": "string",
        },
        "module": Object {
          "$ref": "#/$defs/Module",
        },
      },
      "required": Array [
        "fileName",
        "module",
      ],
      "type": "object",
    },
    "StringType": Object {
      "$id": "#/$defs/StringType",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "maxLength": Object {
          "type": "number",
        },
        "minLength": Object {
          "type": "number",
        },
        "pattern": Object {
          "type": "string",
        },
        "type": Object {
          "const": "string",
        },
      },
      "required": Array [
        "type",
      ],
      "type": "object",
    },
    "TypeInfo": Object {
      "$id": "#/$defs/TypeInfo",
      "oneOf": Array [
        Object {
          "$ref": "#/$defs/ObjectType",
        },
        Object {
          "$ref": "#/$defs/StringType",
        },
        Object {
          "$ref": "#/$defs/BooleanType",
        },
        Object {
          "$ref": "#/$defs/NumberType",
        },
        Object {
          "$ref": "#/$defs/DateType",
        },
        Object {
          "$ref": "#/$defs/NullType",
        },
        Object {
          "$ref": "#/$defs/ArrayType",
        },
        Object {
          "$ref": "#/$defs/TypeReference",
        },
        Object {
          "$ref": "#/$defs/EnumType",
        },
        Object {
          "$ref": "#/$defs/ConstLiteral",
        },
        Object {
          "$ref": "#/$defs/OneOf",
        },
        Object {
          "$ref": "#/$defs/AllOf",
        },
      ],
    },
    "TypeReference": Object {
      "$id": "#/$defs/TypeReference",
      "additionalProperties": undefined,
      "properties": Object {
        "$id": Object {
          "type": "string",
        },
        "$package": Object {
          "type": "string",
        },
        "$ref": Object {
          "type": "string",
        },
      },
      "required": Array [
        "$ref",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema mixed-types 1`] = `
Object {
  "$defs": Object {
    "Location": Object {
      "$id": "#/$defs/Location",
      "additionalProperties": undefined,
      "properties": Object {
        "lat": Object {
          "type": "number",
        },
        "lon": Object {
          "type": "number",
        },
      },
      "required": Array [
        "lat",
        "lon",
      ],
      "type": "object",
    },
    "Person": Object {
      "$id": "#/$defs/Person",
      "additionalProperties": undefined,
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "firstName": Object {
          "type": "string",
        },
        "isActive": Object {
          "type": "boolean",
        },
        "lastName": Object {
          "type": "string",
        },
      },
      "required": Array [
        "firstName",
        "lastName",
        "age",
        "isActive",
      ],
      "type": "object",
    },
    "Role": Object {
      "$id": "#/$defs/Role",
      "oneOf": Array [
        Object {
          "const": "Guest",
        },
        Object {
          "const": "User",
        },
        Object {
          "const": "Admin",
        },
      ],
    },
  },
}
`;

exports[`Process schema one-of 1`] = `
Object {
  "$defs": Object {
    "Circle": Object {
      "$id": "#/$defs/Circle",
      "additionalProperties": undefined,
      "properties": Object {
        "radius": Object {
          "type": "number",
        },
      },
      "required": Array [
        "radius",
      ],
      "type": "object",
    },
    "Picture": Object {
      "$id": "#/$defs/Picture",
      "additionalProperties": undefined,
      "properties": Object {
        "figures": Object {
          "items": Object {
            "$ref": "#/$defs/Shapes",
          },
          "type": "array",
        },
      },
      "required": Array [
        "figures",
      ],
      "type": "object",
    },
    "Shapes": Object {
      "$id": "#/$defs/Shapes",
      "oneOf": Array [
        Object {
          "$ref": "#/$defs/Circle",
        },
        Object {
          "$ref": "#/$defs/Square",
        },
      ],
    },
    "Square": Object {
      "$id": "#/$defs/Square",
      "additionalProperties": undefined,
      "properties": Object {
        "side": Object {
          "type": "number",
        },
      },
      "required": Array [
        "side",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema required-properties 1`] = `
Object {
  "$defs": Object {
    "Person": Object {
      "$id": "#/$defs/Person",
      "additionalProperties": undefined,
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "firstName": Object {
          "type": "string",
        },
        "isActive": Object {
          "type": "boolean",
        },
        "lastName": Object {
          "type": "string",
        },
      },
      "required": Array [
        "firstName",
        "lastName",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema type-alias 1`] = `
Object {
  "$defs": Object {
    "Circle": Object {
      "$id": "#/$defs/Circle",
      "additionalProperties": undefined,
      "properties": Object {
        "radius": Object {
          "type": "number",
        },
      },
      "required": Array [
        "radius",
      ],
      "type": "object",
    },
    "Decorator": Object {
      "$id": "#/$defs/Decorator",
      "additionalProperties": undefined,
      "properties": Object {
        "figure": Object {
          "$ref": "#/$defs/Figure",
        },
      },
      "required": Array [
        "figure",
      ],
      "type": "object",
    },
    "Figure": Object {
      "$id": "#/$defs/Figure",
      "oneOf": Array [
        Object {
          "$ref": "#/$defs/Circle",
        },
        Object {
          "$ref": "#/$defs/Square",
        },
      ],
    },
    "Square": Object {
      "$id": "#/$defs/Square",
      "additionalProperties": undefined,
      "properties": Object {
        "side": Object {
          "type": "number",
        },
      },
      "required": Array [
        "side",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema typeliteral 1`] = `
Object {
  "$defs": Object {
    "Person": Object {
      "$id": "#/$defs/Person",
      "additionalProperties": undefined,
      "properties": Object {
        "firstName": Object {
          "type": "string",
        },
        "lastName": Object {
          "type": "string",
        },
        "location": Object {
          "additionalProperties": undefined,
          "properties": Object {
            "lat": Object {
              "type": "number",
            },
            "lon": Object {
              "type": "number",
            },
          },
          "required": Array [
            "lat",
            "lon",
          ],
          "type": "object",
        },
      },
      "required": Array [
        "firstName",
        "lastName",
        "location",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Process schema typeliteral-argument 1`] = `
Object {
  "$defs": Object {
    "Person": Object {
      "$id": "#/$defs/Person",
      "additionalProperties": undefined,
      "properties": Object {
        "age": Object {
          "maximum": 99,
          "minimum": 1,
          "type": "number",
        },
        "email": Object {
          "pattern": ".*",
          "type": "string",
        },
        "name": Object {
          "type": "string",
        },
      },
      "required": Array [
        "email",
        "age",
        "name",
      ],
      "type": "object",
    },
  },
}
`;
